Migration to Babel7 and @babel/preset-typescript

mistic opened the pull request:
It solves https://github.com/elastic/kibana/issues/33027

That PR migrates the entire Kibana to the new babel7 along with the replace of the typescript compiler with the `@babel/preset-typescript ` in order to generate the javascript code from the typescript files. It will allow us to simplify our current build configuration and our development environment while it will also improve the optimizer performance. 

**Main Changes**

- Migrate packages and dependencies to babel7
- Remove the need for typescript compiler code generation and replace it with @babel/preset-typescript 
- Run all the jest tests with `babel-jest`
- Migrate failing legacy typescript jest tests to use the new mocks pattern
- Drop dependency on `ts-jest, ts-node and ts-loader`
- Apply `babel-plugin-typescript-strip-namespaces` to strip off the namespaces from `/x-pack[\/\\]plugins[\/\\]infra[\/\\].*[\/\\]graphql/` as it is not supported by the babel preset typescript and we are only using it for type definition.

**Caveats with @babel/preset-typescript**

* Does not support [`namespace`][namespace]s. 

  **Workaround**: Move to using [file exports][fm], or migrate to using the `module { }` syntax instead.
  
* Does not support [`const enum`][const_enum]s because those require type information to compile.

  **Workaround**: Remove the `const`, which makes it available at runtime.
  
* Does not support [`export =`][exin] and [`import =`][exin], because those cannot be compiled to ES.next. 

  **Workaround**: Convert to using `export default` and `export const`, and `import x, {y} from "z"`.

**This caveats info was copied from https://babeljs.io/docs/en/babel-plugin-transform-typescript, you can found the complete info and examples there.**
There is also a lot of good information here https://devblogs.microsoft.com/typescript/typescript-and-babel-7/

**New jest test mocks pattern**
As mentioned above a new pattern to declare the `jest mocks` for the Typescript tests was introduced. The main reason for that decision we made is due to the different hoisting applied by `babel` and `typescript`. As a result from this, in order to keep benefiting from the `auto type inference` when testing, and still be able to use `jest.mock` and `jest.doMock`, every jest mocks that needs to reference something from an import or an external variable, should be declared in a file in the same directory as the original jest test and with the following name `{JEST_TEST_FILE_BASE_NAME}.test.mocks.ts` and then should be imported (with or without any important constants) at the top of the test file. A lot of examples could be found into the current PR. In a more structured explanation, the new pattern could be described as follow:

- Each module could provide a standard mock in `mymodule.mock.ts` in case there are other tests that could benefit from using definitions here. This file would not have any jest.mock calls, just dummy objects.

- Each test defines its mocks in `mymodule.test.mocks.ts`. This file could import relevant mocks from the generalised module's mocks file (*.mock.ts) and call jest.mock for each of them. If there is any relevant dummy mock objects to generalise (and to be used by other tests), the dummy objects could be defined directly on this file.

- Each test would import its mocks from the test mocks file `mymodule.test.mocks.ts`. `mymodule.test.ts` has an import like: `import * as Mocks from './mymodule.test.mocks'`, `import { mockX } from './mymodule.test.mocks'` or just `import './mymodule.test.mocks'` if there isn't anything exported to be used.

**Future Needed Follow Ups**
- Migrate tslint to eslint with typescript-eslint plugin
- Migrate kibana packages that are still using `tsc` build steps to use the `@babel/preset-typescript`
- Remove tsc watch from `kbn-pm`
- Build x-pack with `babel`
- Evaluate what to do with `kbn-plugin-helpers`

mistic committed:
chore(NA): first changes on every package.json order to support new babel 7. chore(NA): build for kbn-pm with babel 7.

mistic committed:
chore(NA): patch babel register to load typescrit

mistic committed:
chore(NA): first working version with babel 7 replacing typescript compiler.

mistic committed:
fix(NA): common preset declaration in order to make it work with babel-loader.

mistic committed:
chore(na): organizing babel preset env package json.

mistic committed:
chore(NA): mocha tests enabled.

mistic committed:
fix(NA): typo on importing

mistic committed:
test(NA): start fixing tests.

mistic committed:
test(NA): majority of x-pack tests ported to use babel-jest

mistic committed:
fix(NA): report info button test with babel-jest.

mistic committed:
fix(NA): polling service tests.

mistic committed:
test(na): fix server plugins plugin tests.

mistic committed:
test(NA): batch of test fixs for jest tests under babel-jest hoisting.

mistic committed:
chore(NA): add babel plugin to hoist mock prefixed vars on jest tests.

mistic committed:
chore(NA): update yarn.lock file.

mistic committed:
chore(NA): tests passing.

mistic committed:
chore(NA): remove wrong dep

mistic committed:
chore(NA): fix tsconfig

mistic committed:
chore(NA): merge and solve conflicts with master.

mistic committed:
chore(NA): merge and solve conflicts with master.

mistic committed:
chore(NA): skip babel for ts-jest.

mistic committed:
chore(NA): selectively apply the plugin to strip off namespace from ts files.

mistic committed:
chore(NA): remove not needed changes from ts tests

mistic committed:
chore(NA): removed ts-jest dependency. chore(NA): migrate ts tests on x-pack to use babel-jest with the new pattern.

mistic committed:
chore(NA): migrate kibana default distribution typescript tests to run with babel-jest and the new test mock pattern.

mistic committed:
Merge branch 'master' into migrate-to-babel-7

mistic committed:
chore(NA): merge and solve conflicts with master.

elasticmachine commented:
Pinging @elastic/kibana-operations

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5099/)

sqren commented on x-pack/plugins/apm/public/components/shared/PropertiesTable/__test__/tabConfig.test.tsx:
```diff
@@ -4,18 +4,33 @@
  * you may not use this file except in compliance with the Elastic License.
  */
 
+jest.mock('../tabConfigConst', () => {
```
Thanks for cleaning this up 👍

sorenlouv reviewed (approved):
APM changes LGTM

monfera commented:
Aren't we on `typescript` `v3.3.3333` since? Would it be worth rebasing on `master`? (not sure if it's needed, but since TS seems central to the change, maybe)

Also, is CI in red due to flakiness or something else?

mistic committed:
chore(NA): fix problems reported by eslint

mistic committed:
Merge branch 'master' into migrate-to-babel-7

monfera reviewed (approved):
Looks great, Canvas is functional!

Besides checking basic functionality, I also checked if:
- the [type strength PR](https://github.com/elastic/kibana/pull/32905) is still doing its thing (yes)
- there's a change to what runtime functions get compiled in, which may worsen performance (yes there's a change, instead of the `__read` function now it's a `_slicedToArray` function which is *way* smarter and just lets through the value, no useless recreation of the array; interestingly the FPS with the 100 element stress test went from 30FPS to 27FPS but maybe it's unrelated to the branch)

cc @clintandrewhall

walterra reviewed (approved):
ML changes LGTM

mistic committed:
chore(NA): fix license ovveride for babel-plugin-mock-imports

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5236/)

mistic committed:
chore(NA): update jest integration tests for kbn pm

mistic committed:
Merge branch 'master' into migrate-to-babel-7

mistic committed:
chore(NA): update babel jest integration tests for kbn pm.

spalger commented on src/core/public/core_system.test.ts:
```diff
@@ -17,119 +17,23 @@
+  MockNotificationsService,
+  mockNotificationStart,
+  MockUiSettingsService,
+} from './core_system.mock';
```
Can we use `.mocks.ts` instead since there are multiple mocks?

mistic replied:
We can @spalger . The idea behind `.mock.ts` was basically to follow the same pattern we have for the test jest files' extensions `.test.ts` which are singular but also contains multiple tests. Do you still think we should use `.mocks.ts` in plural?

joshdover replied:
I think using `.mocks.ts` may be confusing in conjunction with the [the convention that's being established](https://github.com/elastic/kibana/issues/31486) for modules providing generic mocks that can be used in all tests. I think a better approach would be:
- Each module provides a standard mock in `mymodule.mock.ts`. This file would not have any `jest.mock` calls, just dummy objects that other tests can use.
- Each test defines its mocks in `mymodule.test.mocks.ts`. This file would import the relevant mocks needed from the module's dependencies (`*.mock.ts`) and call `jest.mock` for each of them.
- Each test would import its mocks from the mocks file. `mymodule.test.ts` has an import like: `import * as Mocks from './mymodule.test.mocks`.

mistic replied:
@joshdover I think it is a really good idea! I was not aware of that new convention by the platform team, thanks for letting me know!

mistic replied:
@spalger @joshdover I just have migrated all the jest mocks to `.test.mocks.js`

spalger commented on src/dev/jest/ts_babel_jest_disable_hoist.js:
```diff
@@ -0,0 +1,56 @@
+/*
```
Should probably delete this module.

mistic replied:
👍

spalger commented on src/optimize/base_optimizer.js:
```diff
@@ -363,7 +358,7 @@ export default class BaseOptimizer {
           },
           {
-            resource: createSourceFileResourceSelector(/\.js$/),
+            resource: createSourceFileResourceSelector(/\.(js|jsx|tsx|ts)$/),
```
We don't need to support .jsx

mistic replied:
👍

spalger commented on src/optimize/base_optimizer.js:
```diff
@@ -389,7 +384,7 @@ export default class BaseOptimizer {
 
       resolve: {
-        extensions: ['.js', '.json'],
+        extensions: ['.js', '.jsx', '.ts', '.tsx', '.json'],
```
Adding `.jsx` causes webpack to do extra lookups that are not necessary since use of `.jsx` is super uncommon and we haven't supported it for a while.

mistic replied:
👍

spalger commented on src/setup_node_env/babel_register/register.js:
```diff
@@ -66,10 +66,11 @@ if (global.__BUILT_WITH_BABEL__) {
   presets: [
     require.resolve('@kbn/babel-preset/node_preset')
   ],
+  extensions: ['.es6', '.es', '.jsx', '.js', '.mjs', '.ts', '.tsx'],
```
lots of extra extensions here, lets limit this to `'.js', '.ts', '.tsx'`

mistic replied:
👍

spalger commented on src/dev/jest/config.js:
```diff
@@ -82,8 +77,7 @@ export default {
   transform: {
-    '^.+\\.js$': '<rootDir>/src/dev/jest/babel_transform.js',
-    '^.+\\.tsx?$': '<rootDir>/src/dev/jest/ts_transform.js',
+    '^.+\\.(jsx?|tsx?)$': '<rootDir>/src/dev/jest/babel_transform.js',
```
There are a lot of jsx mentions, we don't need to worry about that

mistic replied:
👍

spalger reviewed (commented):
Looking great!

mistic committed:
test(NA): update jest integration snapshot for kbn pm.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5268/)

mistic committed:
chore(NA): apply changes according to the pull request reviews.

mistic committed:
chore(NA): apply changes according to the pull request reviews.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5283/)

mistic commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5290/)

mistic committed:
refact(NA): migrate jest tests to the new pattern.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5351/)

mistic committed:
fix(NA): babel 7 polyfill in the tests bundle.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5455/)

mistic committed:
chore(NA): restore needed step in order to compile x-pack with typescript.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5483/)

mistic committed:
chore(NA): change build to compile typescript with babel for the oss code. chore(NA): change transpile typescript task to only transpile types for x-pack. refact(NA): common preset for babel 7

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5534/)

mistic committed:
Revert "chore(NA): change build to compile typescript with babel for the oss code. chore(NA): change transpile typescript task to only transpile types for x-pack. refact(NA): common preset for babel 7"

This reverts commit 2707d538f5126c582a00580a8489361592c80d70.

mistic committed:
chore(NA): merge and solve conflicts with master.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5632/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5642/)

mistic committed:
Merge branch 'master' into migrate-to-babel-7

mistic committed:
fix(NA): import paths for tabConfigConst

mistic committed:
chore(NA): fix transpiling error on browser tests

mistic committed:
chore(NA): simplify kbn babel preset package.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5716/)

mistic committed:
chore(NA): migrate build to use babel transpiler for typescript excluding xpack.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5733/)

mistic commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5734/)

mistic committed:
fix(NA): introduced error on test quick task.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5735/)

mistic committed:
fix(NA): fix preset for client side code on build.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5741/)

mistic committed:
fix(NA): build with babel

mistic committed:
fix(NA): negated patterns in the end.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5753/)

mistic committed:
fix(NA): kbn_tp_sample_panel_action creation.

mistic committed:
chore(NA): merge and solve conflicts with last master.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5849/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5858/)

mistic committed:
fix(NA): babel typescript transform plugin workaround when exporting interface name.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5868/)

joshdover commented on src/core/server/elasticsearch/cluster_client.test.mocks.ts:
```diff
@@ -0,0 +1,35 @@
+export const MockClient = jest.fn();
+jest.mock('elasticsearch', () => ({
+  // Jest types don't include `requireActual` right now.
+  ...(jest as any).requireActual('elasticsearch'),
```
This was fixed when we upgraded to jest recently, so you should be able to remove the cast to `any` and the comment here.

mistic replied:
👍

joshdover commented on src/core/server/plugins/plugin.test.ts:
```diff
@@ -29,15 +29,15 @@ import { createPluginInitializerContext, createPluginStartContext } from './plug
   virtual: true,
 });
-jest.mock(join('plugin-with-wrong-initializer-path', 'server'), () => ({ plugin: {} }), {
+jest.doMock(join('plugin-with-wrong-initializer-path', 'server'), () => ({ plugin: {} }), {
```
Could we use doMock everywhere instead of the new `*.test.mocks.ts` files?

mistic replied:
We really need that pattern. Typescript has an open bug which is keep allowing us to build the tests as we are doing with the ts-jest https://github.com/Microsoft/TypeScript/issues/16166#issue-232515907

As the imports are not being hoisted, things like the following would work on `ts-jest` but not with `babel-jest`

```
const mockJobQueueClient = { getInfo: jest.fn() };
jest.doMock('../lib/job_queue_client', () => ({ jobQueueClient: mockJobQueueClient }));

import React from 'react';
import { mountWithIntl } from 'test_utils/enzyme_helpers';
import { ReportInfoButton } from './report_info_button';
```

So we need that pattern to enclose the jest mocks in an import that would also be hoisted on the top with `babel-jest` before the other imports

joshdover commented on src/core/server/root/index.test.mocks.ts:
```diff
@@ -0,0 +1,33 @@
+}));
+
+export const mockServer = { start: jest.fn(), stop: jest.fn() };
+jest.mock('../', () => ({ Server: jest.fn(() => mockServer) }));
```
Why is this one `mock` and the other two are `doMock`?

mistic replied:
I just touched the tests that were failing, for example on this one, we can keep using `jest.mock` as the variables we are linking are also prefixed with `mock`

joshdover reviewed (approved):
@mistic Answered my questions offline, LGTM. I think it would be good if some docs were added to either CONTRIBUTING.md or the "Core Development" section in the elastic.co docs.

mistic committed:
chore(NA): merge and solve conflicts with master.

mistic committed:
refact(NA): remove not needed type cast to any on jest test.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6238/)

mistic committed:
docs(NA): add developement documentation about jest mocks test pattern.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6290/)

spalger reviewed (approved):
LGTM

mistic committed:
chore(NA): merge and solve conflicts with last master.

mistic committed:
chore(NA): missing unmerged path.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6740/)

mistic committed:
chore(NA): fix jest tests for template.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6743/)

loixlab committed:
[CCR] Client integration tests (table lists) (#33525)

azasypkin committed:
Force user to re-authenticate if token refresh fails with `400` status code. (#33774)

ycombinator committed:
Improve performance of the Logstash Pipeline Viewer (#33793)

Resolves #27513.

_This PR is a combination of #31293 (the code changes) + #33570 (test updates). These two PRs were individually reviewed and merged into a feature branch. This combo PR here simply sets up the merge from the feature branch to `master`._

Summary of changes, taken from #31293:

## Before this PR
The Logstash Pipeline Viewer UI would make a single Kibana API call to fetch all the information necessary to render the Logstash pipeline. This included information necessary to render the detail drawer that opens up when a user clicks on an individual vertex in the pipeline.

Naturally, this single API call fetched _a lot_ of data, not just from the Kibana server but also, in turn, from Elasticsearch as well. The "pro" of this approach was that the user would see instantaneous results if they clicked on a vertex in a pipeline and opened the detail drawer for that vertex. The "cons" were the amount of computation Elasticsearch had to perform and the amount of data being transferred over the wire between Elasticsearch and the Kibana server as well as between the Kibana server and the browser.

## With this PR
This PR makes the Kibana API call to fetch data necessary for **initially** rendering the pipeline — that is, with the detail drawer closed — much lighter. When the user clicks on a vertex in a pipeline, a second API call is then made to fetch data necessary for the detail drawer.

## Gains, by the numbers

Based on a simple, 1-input, 1-filter, and 1-output pipeline.

* Before this PR, the Elasticsearch `logstash_stats` API responses (multiple calls were made using the `composite` aggregation over the `date_histogram` aggregation) generated a total of 1228 aggregation buckets (before any `filter_path`s were applied but across all `composite` "pages"). With this PR, the single `logstash_stats` API response (note that this is just for the initial rendering of the pipeline, with the detail drawer closed) generated 12 buckets (also before any `filter_path`s were applied). That's a **99.02% reduction** in number of buckets.

* Before this PR, the Elasticsearch `logstash_stats` API responses added up to 70319 bytes. With this PR, the single `logstash_stats` API response for the same pipeline is 746 bytes. That's a **98.93% reduction** in size.

* Before this PR, the Elasticsearch `logstash_state` API response was 7718 bytes. With this PR, the API response for the same pipeline is 2328 bytes. That's a **69.83% reduction** in size.

* Before this PR the Kibana API response was 51777 bytes. With this PR, the API response for the same pipeline is 2567 bytes (again, note that this is just for the initial rendering of the pipeline, with the detail drawer closed). That's a **95.04% reduction** in size.

nreese committed:
[Maps] split settings into layer and source panels (#33788)

* [Maps] split settings into layer and source panels

* fix SCSS import

jbudz committed:
[env] exit if starting as root (#21563)

* [env] exit if starting as root

* fix windows

* s/--allow-root

Commit:
Typescript sample panel action (#33602)

* Typescript sample panel action

* Update EUI version to match main cabana version

* update yarn.lock

* add back typings include

* use correct relative path

PhilippBaranovskiy committed:
Home page "recent links" should communicate saved object type #21896 (#33694)

* adds object type for screen order
* adds object type for pointer hovering
* Update src/legacy/ui/public/chrome/directives/header_global_nav/components/header.tsx

Co-Authored-By: rockfield <philipp.b@ya.ru>

mistic committed:
chore(NA): merge and solve conflicts with last master.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6853/)

snide reviewed (approved):
Reviewed the .scss related files that triggered the design-team review. Everything looks in order.

mistic commented:
7.x: pending
